src/main/scala/com/fulcrumgenomics/bam/SplitBam.scala

```diff
@@ -55,77 +55,101 @@ object SplitType extends FgBioEnum[SplitType] {
+      info.writer.close()
+      logger.info(f"Wrote ${info.count}%,d records to ${info.bam.getFileName}")
     }
 
-    if (counter.countOf(unknownBamAndWriter) == 0) {
+    if (unknownBamAndWriter.count == 0) {
       Files.delete(unknownBamAndWriter.bam)
     }
   }
 
-  /** Stores the path to the BAM and the associated writer */
-  private case class WriterInfo(name: String, bam: PathToBam, writer: SamWriter) extends Closeable{
-    def close(): Unit = writer.close()
+  /** Gets the output path for the writer with a given name. */
+  private[bam] def toOutput(name: String): PathToBam = {
+    val outputDir = output.getParent
+    val prefix    = output.getFileName
+    outputDir.resolve(PathUtil.sanitizeFileName(s"$prefix.$name.bam"))
   }
 
   /** Initializes the writers and returns a map from each read group to a writer. */
   private def createWriters(header: SAMFileHeader, splitBy: SplitType): Map[SAMReadGroupRecord, WriterInfo] = {
-    splitBy match {
-      case SplitType.Library =>
-        header.getReadGroups.toSeq.groupBy { rg => rg.getLibrary }
-          .flatMap { case (library, readGroups) =>
-            val bam = toOutput(library)
-            val writer = SamWriter(bam, header)
-            readGroups.map { rg => rg -> WriterInfo(name=library, bam=bam, writer=writer) }
-          }
-      case SplitType.ReadGroup =>
-        header.getReadGroups.map { rg =>
-          val bam = toOutput(rg.getId)
-          val writer = SamWriter(bam, header)
-          rg -> WriterInfo(name=rg.getId, bam=bam, writer=writer)
-        }.toMap
+    val groups: Map[String, Seq[SAMReadGroupRecord]]  = splitBy match {
+      case SplitType.Library   => header.getReadGroups.toSeq.groupBy { rg => rg.getLibrary }
+      case SplitType.ReadGroup => header.getReadGroups.map(rg => rg.getId -> Seq(rg)).toMap
+    }
+    logger.info(f"Outputting to ${groups.size}%,d files ($splitBy).")
+    groups.flatMap { case (library, readGroups) =>
+      val bam    = toOutput(library)
+      val writer = SamWriter(
+        path   = bam,
+        header = toOutputHeader(header=header, readGroups:_*),
+        async  = !noAsyncWriting
+      )
+      readGroups.map { rg => rg -> WriterInfo(name=library, bam=bam, writer=writer) }
     }
   }
 
-  private[bam] def toOutput(name: String): PathToBam = {
-    val outputDir = output.getParent
-    val prefix    = output.getFileName
-    outputDir.resolve(PathUtil.sanitizeFileName(s"$prefix.$name.bam"))
+  /** Stores the path to the BAM and the associated writer */
+  private case class WriterInfo(name: String, bam: PathToBam, writer: Writer[SamRecord], var count: Long = 0) extends Closeable {
+    def close(): Unit = writer.close()
+  }
+
+  /** Efficiently creates an output header with a single read group from the input header.
+    *
+    * Developer note: `SamFileHeader.clone` is very memory inefficient.  It creates a `String` from the header to clone,
+    * the decodes it!
```

Comment:
Imagine having to parse the header for 10K `SamWriters`!  This takes forever.

Reply:
Note that previously the output would contain all the read groups in the header.  In this version, only the single read group being output is added to the output header.